File: hw3/hw3.py
def hw3(input, output):
        fInput = open(input,'r')
        f = open(output,'w')
        stack = []
        map1 = {}
        for line in fInput:
                if line[0].isalpha():
                        parsePrimitive(line, stack, f, map1)
                elif line[0] == ':':
                        parseBooleanOrError(line, stack)
        fInput.close()


def parsePrimitive(line, stack, f, map1):
        if line.startswith('add'):
                doAdd(stack, map1)
        elif line.startswith('sub'):
                doSub(stack, map1)
        elif line.startswith('mul'):
                doMul(stack, map1)
        elif line.startswith('div'):
                doDiv(stack, map1)
        elif line.startswith('rem'):
                doRem(stack, map1)
        elif line.startswith('pop'):
                doPop(stack, map1)
        elif line.startswith('push'):
                doPush(stack, line)
        elif line.startswith('swap'):
                doSwap(stack, map1)
        elif line.startswith('neg'):
                doNeg(stack, map1)
        elif line.startswith('quit'):
                doQuit(stack, f)
        elif line.startswith('and'):
                doAnd(stack)
        elif line.startswith('or'):
                doOr(stack)
        elif line.startswith('not'):
                doNot(stack)
        elif line.startswith('equal'):
                doEqual(stack)
        elif line.startswith('lessThan'):
                doLessThan(stack)
        elif line.startswith('bind'):
                doBind(stack, map1)
        elif line.startswith('if'):
                doIf(stack)

def doAdd(stack, map1):
        if len(stack) < 2:
                return stack.insert(0, ':error:')
        elif stack[0][0] == ':' or stack[1][0] == ':':
                return stack.insert(0, ':error:')
        elif stack[0] in map1 and stack[1] in map1:
                xValue = map1[stack[1]]
                x = int(xValue)
                yValue = map1[stack[0]]
                y = int(yValue)
                stack.pop(0)
                stack.pop(0)
                newTop = x+y
                return stack.insert(0, str(newTop))
        elif stack[0] in map1:
                xValue = map1[stack[1]]
                x = int(xValue)
                y = int(stack[0])
                stack.pop(0)
                stack.pop(0)
                newTop = x+y
                return stack.insert(0, str(newTop))
        elif stack[1] in map1:
                yValue = map1[stack[0]]
                y = int(yValue)
                x = int(stack[1])
                stack.pop(0)
                stack.pop(0)
                newTop = x+y
                return stack.insert(0, str(newTop))
        else:
                x = int(stack[1])
                y = int(stack[0])
                stack.pop(0)
                stack.pop(0)
                newTop = x+y
                return stack.insert(0, str(newTop))


def doSub(stack, map1):
        if len(stack) < 2:
                return stack.insert(0, ':error:')
        elif stack[0][0] == ':' or stack[1][0] == ':':
                return stack.insert(0, ':error:')
        elif stack[0] in map1 and stack[1] in map1:
                xValue = map1[stack[1]]
                x = int(xValue)
                yValue = map1[stack[0]]
                y = int(yValue)
                stack.pop(0)
                stack.pop(0)
                newTop = x-y
                return stack.insert(0, str(newTop))
        elif stack[0] in map1:
                xValue = map1[stack[1]]
                x = int(xValue)
                y = int(stack[0])
                stack.pop(0)
                stack.pop(0)
                newTop = x-y
                return stack.insert(0, str(newTop))
        elif stack[1] in map1:
                yValue = map1[stack[0]]
                y = int(yValue)
                x = int(stack[1])
                stack.pop(0)
                stack.pop(0)
                newTop = x-y
                return stack.insert(0, str(newTop))
        else:
                x = int(stack[1])
                y = int(stack[0])
                stack.pop(0)
                stack.pop(0)
                newTop = x-y
                return stack.insert(0, str(newTop))



def doMul(stack, map1):
        if len(stack) < 2:
                return stack.insert(0, ':error:')
        elif stack[0][0] == ':' or stack[1][0] == ':':
                return stack.insert(0, ':error:')
        elif stack[0] in map1 and stack[1] in map1:
                xValue = map1[stack[1]]
                x = int(xValue)
                yValue = map1[stack[0]]
                y = int(yValue)
                stack.pop(0)
                stack.pop(0)
                newTop = x*y
                return stack.insert(0, str(newTop))
        elif stack[0] in map1:
                xValue = map1[stack[1]]
                x = int(xValue)
                y = int(stack[0])
                stack.pop(0)
                stack.pop(0)
                newTop = x*y
                return stack.insert(0, str(newTop))
        elif stack[1] in map1:
                yValue = map1[stack[1]]
                y = int(yValue)
                x = int(stack[0])
                stack.pop(0)
                stack.pop(0)
                newTop = x*y
                return stack.insert(0, str(newTop))
        else:
                x = int(stack[1])
                y = int(stack[0])
                stack.pop(0)
                stack.pop(0)
                newTop = x*y
                return stack.insert(0, str(newTop))



def doDiv(stack, map1):
        if len(stack) < 2:
                return stack.insert(0, ':error:')
        elif stack[0][0] == ':' or stack[1][0] == ':':
                return stack.insert(0, ':error:')
        elif stack[0] in map1 and stack[1] in map1:
                xValue = map1[stack[1]]
                x = int(xValue)
                yValue = map1[stack[0]]
                y = int(yValue)
                stack.pop(0)
                stack.pop(0)
                newTop = x//y
                return stack.insert(0, str(newTop))
        elif stack[0] in map1:
                xValue = map1[stack[1]]
                x = int(xValue)
                y = int(stack[0])
                stack.pop(0)
                stack.pop(0)
                newTop = x//y
                return stack.insert(0, str(newTop))
        elif stack[1] in map1:
                yValue = map1[stack[0]]
                y = int(yValue)
                x = int(stack[1])
                stack.pop(0)
                stack.pop(0)
                newTop = x//y
                return stack.insert(0, str(newTop))
        else:
                x = int(stack[1])
                y = int(stack[0])
                if y == 0:
                        return stack.insert(0, ':error:')
                else:
                        stack.pop(0)
                        stack.pop(0)
                        newTop = x//y
                        return stack.insert(0, str(newTop))



def doRem(stack, map1):
        if len(stack) < 2:
                return stack.insert(0, ':error:')
        elif stack[0][0] == ':' or stack[1][0] == ':':
                return stack.insert(0, ':error:')
        elif stack[0] in map1 and stack[1] in map1:
                xValue = map1[stack[1]]
                x = int(xValue)
                yValue = map1[stack[0]]
                y = int(yValue)
                stack.pop(0)
                stack.pop(0)
                newTop = x % y
                return stack.insert(0, str(newTop))
        elif stack[0] in map1:
                xValue = map1[stack[1]]
                x = int(xValue)
                y = int(stack[0])
                stack.pop(0)
                stack.pop(0)
                newTop = x % y
                return stack.insert(0, str(newTop))
        elif stack[1] in map1:
                yValue = map1[stack[0]]
                y = int(yValue)
                x = int(stack[1])
                stack.pop(0)
                stack.pop(0)
                newTop = x % y
                return stack.insert(0, str(newTop))
        else:
                x = int(stack[1])
                y = int(stack[0])
                if y == 0:
                        return stack.insert(0, ':error:')
                else:
                        stack.pop(0)
                        stack.pop(0)
                        newTop = x % y
                        return stack.insert(0, str(newTop))


def doPop(stack):
	if len(stack) < 1:
		return stack.insert(0, ':error:')
	else:
		return stack.pop(0)


def doPush(stack, line):
        getList = line.split()
        if getList[1][0] == '-':
                if getList[1][1:] == '0':
                        return stack.insert(0,'0')
                elif getList[1][1:].isdigit():
                        return stack.insert(0, getList[1])
                else:
                        return stack.insert(0, ':error:')
        elif getList[1].isdigit():
                return stack.insert(0, getList[1])
        elif getList[1].startswith("\"") and getList[1].endswith("\""):
                noQuotes = getList[1][1:-1]
                return stack.insert(0, noQuotes)
        elif getList[1][0].isalpha():
                return stack.insert(0, getList[1])
        else:
                return stack.insert(0, ':error:')


def doSwap(stack):
	if len(stack) < 2:
		return stack.insert(0, ':error:')
	else:
		x = stack[1]
		y = stack[0]
		stack.pop(0)
		stack.pop(0)
		stack.insert(0, y)
		return stack.insert(0, x)


def doNeg(stack):
	if len(stack) < 1:
		return stack.insert(0, ':error:')
	elif stack[0][0] == ':':
		return stack.insert(0, ':error:')
	else:
		x = int(stack[0])
		stack.pop(0)
		newTop = -1*x
		return stack.insert(0, str(newTop))


def doQuit(stack, f):
	for ele in stack:
		f.write(ele + '\n')
	f.close()


def doAnd(stack):
        if len(stack) < 2:
                return stack.insert(0, ':error:')
        elif (stack[1] == ":true:" or stack[1] == ":false:") and (stack[0] == ":true:" or stack[0] == ":false:"):
                x = stack[1]
                y = stack[0]
                stack.pop(0)
                stack.pop(0)
                if x == ":true:":
                        x = True
                if y == ":true:":
                        y = True
                if x == ":false:":
                        x = False
                if y == ":false:":
                        y = False
                newTop = x and y
                if newTop == True:
                        newTop = ":true:"
                if newTop == False:
                        newTop = ":false:"
                return stack.insert(0, newTop)
        else:
                return stack.insert(0, ':error:')


def doOr(stack):
        if len(stack) < 2:
                return stack.insert(0, ':error:')
        elif (stack[1] == ":true:" or stack[1] == ":false:") and (stack[0] == ":true:" or stack[0] == ":false:"):
                x = stack[1]
                y = stack[0]
                stack.pop(0)
                stack.pop(0)
                if x == ":true:":
                        x = True
                if y == ":true:":
                        y = True
                if x == ":false:":
                        x = False
                if y == ":false:":
                        y = False
                newTop = x or y
                if newTop == True:
                        newTop = ":true:"
                if newTop == False:
                        newTop = ":false:"
                return stack.insert(0, newTop)
        else:
                return stack.insert(0, ':error:')

def doNot(stack):
        if len(stack) < 1:
                return stack.insert(0, ':error:')
        elif stack[0] == ":true:" or stack[0] == ":false:":
                x = stack[0]
                stack.pop(0)
                if x == ":true:":
                        x = True
                if x == ":false:":
                        x = False
                newTop = not x
                if newTop == True:
                        newTop = ":true:"
                if newTop == False:
                        newTop = ":false:"
                return stack.insert(0, newTop)
        else:
                return stack.insert(0, ':error:')


def doEqual(stack):
        if len(stack) < 2:
                return stack.insert(0, ':error:')
        elif isinstance(int(stack[0]), int) and isinstance(int(stack[1]), int):
                x = int(stack[1])
                y = int(stack[0])
                stack.pop(0)
                stack.pop(0)
                newTop = (x == y)
                if newTop == True:
                        newTop = ":true:"
                if newTop == False:
                        newTop = ":false:"
                return stack.insert(0, newTop)
        else:
                return stack.insert(0, ':error:')


def doLessThan(stack):
        if len(stack) < 2:
                return stack.insert(0, ':error:')
        elif isinstance(int(stack[0]), int) and isinstance(int(stack[1]), int):
                x = int(stack[1])
                y = int(stack[0])
                stack.pop(0)
                stack.pop(0)
                newTop = (x < y)
                if newTop == True:
                        newTop = ":true:"
                if newTop == False:
                        newTop = ":false:"
                return stack.insert(0, newTop)
        else:
                return stack.insert(0, ':error:')


def doBind(stack, map1):
        if len(stack) < 1:
                return stack.insert(0, ':error:')
        else:
                x = stack[1]
                y = stack[0]
                stack.pop(0)
                stack.pop(0)
                map1[x] = y
                return stack.insert(0, ':unit:')
                
def doIf(stack):
        if len(stack) < 3:
                return stack.insert(0, ':error:')
        elif stack[2] == ":true:" or stack[2] == ":false:":
                z = stack[2]
                y = stack[1]
                x = stack[0]
                stack.pop(0)
                stack.pop(0)
                stack.pop(0)
                if z == ":true:":
                        return stack.insert(0, x)
                if z == ":false:":
                        return stack.insert(0,y)
        else:
                return stack.insert(0, ':error:')

def parseBooleanOrError(line, stack):
	if line[1] == 'e':
		return stack.insert(0,':error:')
	elif line[1] == 't':
		return stack.insert(0,':true:')
	else:
		return stack.insert(0,':false:')


hw3("input_9.txt","output_9.txt")

File: hw3/src/hw3.java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.io.PrintStream;
import java.util.ArrayList;
import java.util.Arrays;

public class hw3 {
	public static void hw3(String input, String output) throws IOException{
	
		PrintStream myconsole=new PrintStream(new File(output));
		System.setOut(myconsole);
		
		BufferedReader in = new BufferedReader(new FileReader(input));
		
		ArrayList stack = new ArrayList();
		String line = in.readLine();
		while(line!=null){
			
		
			if(Character.isLetter(line.charAt(0))){
				stack = parsePrimitive(line, stack, myconsole);
			}
			else if(line.charAt(0)==':'){
				stack = parseBooleanOrError(line, stack);
			}
			else{
				myconsole.println("Error command!");
			}
			
			line = in.readLine();
		
		}
		in.close();
		
		
	}
	
	public static ArrayList parseBooleanOrError(String line, ArrayList stack) {
		if (line.startsWith(":e")){
			stack.add(0, ":error:");
		}
		else if (line.startsWith(":t")){
			stack.add(0, ":true:");
		}
		else if (line.startsWith(":f")){
			stack.add(0, ":false:");
		}
		
		return stack;
	}

	public static ArrayList doMul(ArrayList stack) {
		if (stack.size()<2){
			stack.add(0, ":error:");
		}
		else if (((String) stack.get(0)).charAt(0) == ':' || ((String) stack.get(1)).charAt(0) == ':'){
			stack.add(0, ":error:");
		}
		else{
			int x = Integer.parseInt((String) stack.get(1));
			int y = Integer.parseInt((String) stack.get(0));
			stack.remove(0);
			stack.remove(0);
			Integer newTop = x*y;
			stack.add(0, newTop.toString());
		}
		return stack;
	}

	public static ArrayList doSub(ArrayList stack) {
		if (stack.size()<2){
			stack.add(0, ":error:");
		}
		else if (((String) stack.get(0)).charAt(0) == ':' || ((String) stack.get(1)).charAt(0) == ':'){
			stack.add(0, ":error:");
		}
		else{
			int x = Integer.parseInt((String) stack.get(1));
			int y = Integer.parseInt((String) stack.get(0));
			stack.remove(0);
			stack.remove(0);
			Integer newTop = x-y;
			stack.add(0, newTop.toString());
		}
		return stack;
	}

	public static ArrayList doAdd(ArrayList stack) {
		if (stack.size()<2){
			stack.add(0, ":error:");
		}
		else if (((String) stack.get(0)).charAt(0) == ':' || ((String) stack.get(1)).charAt(0) == ':'){
			stack.add(0, ":error:");
		}
		else{
			int x = Integer.parseInt((String) stack.get(1));
			int y = Integer.parseInt((String) stack.get(0));
			stack.remove(0);
			stack.remove(0);
			Integer newTop = x+y;
			stack.add(0, newTop.toString());
		}
		return stack;
	}

	public static ArrayList parsePrimitive(String line, ArrayList stack, PrintStream myconsole){
		if (line.startsWith("add")){
			stack = doAdd(stack);
		}
		else if (line.startsWith("sub")){
			stack = doSub(stack);
		}
		else if (line.startsWith("mul")){
			stack = doMul(stack);
		}
		else if (line.startsWith("div")){
			stack = doDiv(stack);
		}
		else if (line.startsWith("rem")){
			stack = doRem(stack);
		}
		else if (line.startsWith("pop")){
			stack = doPop(stack);
		}
		else if (line.startsWith("push")){
			stack = doPush(stack, line);
		}
		else if (line.startsWith("swap")){
			stack = doSwap(stack);
		}
		else if (line.startsWith("neg")){
			stack = doNeg(stack);
		}
		else if (line.startsWith("quit")){
			doQuit(stack, myconsole);
		}
		else if (line.startsWith("and")){
			stack = doAnd(stack);
		}
		else if (line.startsWith("or")){
			stack = doOr(stack);
		}
		else if (line.startsWith("equal")){
			stack = doEqual(stack);
		}
		else if (line.startsWith("lessThan")){
			stack = doLessThan(stack);
		}
		else if (line.startsWith("bind")){
			stack = doBind(stack);
		}
		else if (line.startsWith("if")){
			stack = doIf(stack);
		}
		
		return stack;
	}

	public static void doQuit(ArrayList stack, PrintStream myconsole) {
		for (int i = 0; i < stack.size(); i++){
			myconsole.println(stack.get(i));
		}
		myconsole.close();
	}

	public static ArrayList doNeg(ArrayList stack) {
		if (stack.isEmpty()){
			stack.add(0, ":error:");
		}
		else if (((String) stack.get(0)).charAt(0) == ':'){
			stack.add(0, ":error:");
		}
		else{
			int x = Integer.parseInt((String) stack.get(0));
			Integer newTop = -1*x;
			stack.remove(0);
			stack.add(0, newTop.toString());
			}
		return stack;
	}

	private static ArrayList doSwap(ArrayList stack) {
		if (stack.size() < 2){
			stack.add(0, ":error:");
		}
		else{
			String x = (String) stack.get(1);
			String y = (String) stack.get(0);
			stack.remove(0);
			stack.remove(0);
			stack.add(0, y);
			stack.add(0, x);
		}
		return stack;
	}

	public static ArrayList doPush(ArrayList stack, String line) {
//		String[] lineList = line.split(" ");
//		String getNum = Arrays.asList(lineList).get(1);
//		stack.add(0, getNum);
		
		//stack.add(0, line.substring(5));
		String getNum = line.substring(5);
		System.out.println(getNum);
		if (getNum.charAt(0) == '-'){
			if (getNum.substring(1).equals("0")){
				stack.add("0");
			}
			else if (getNum.substring(1).matches("[0-9]+")){
				stack.add(0, getNum);
			}
			else{
				stack.add(0, ":error:");
			}
		}
		else if (getNum.matches("[0-9]+")){
			stack.add(0, getNum);
		}
		else if (getNum.matches("[a-zA-Z]+")){
			stack.add(0, getNum);
		}
		else{
			stack.add(0, ":error:");
		}
		return stack;
	}

	public static ArrayList doPop(ArrayList stack) {
		if (stack.size() < 1){
			stack.add(0, ":error:");
		}
		else{
			stack.remove(0);
		}
		return stack;
	}

	public static ArrayList doRem(ArrayList stack) {
		if (stack.size()<2){
			stack.add(0, ":error:");
		}
		else if (((String) stack.get(0)).charAt(0) == ':' || ((String) stack.get(1)).charAt(0) == ':'){
			stack.add(0, ":error:");
		}
		else{
			int x = Integer.parseInt((String) stack.get(1));
			int y = Integer.parseInt((String) stack.get(0));
			if (y == 0){
				stack.add(0, ":error:");
			}
			else{
				stack.remove(0);
				stack.remove(0);
				Integer newTop = x%y;
				stack.add(0, newTop.toString());
			}
		}
		return stack;
	}

	public static ArrayList doDiv(ArrayList stack) {
		if (stack.size()<2){
			stack.add(0, ":error:");
		}
		else if (((String) stack.get(0)).charAt(0) == ':' || ((String) stack.get(1)).charAt(0) == ':'){
			stack.add(0, ":error:");
		}
		else{
			int x = Integer.parseInt((String) stack.get(1));
			int y = Integer.parseInt((String) stack.get(0));
			if (y == 0){
				stack.add(0, ":error:");
			}
			else{
				stack.remove(0);
				stack.remove(0);
				Integer newTop = x/y;
				stack.add(0, newTop.toString());
			}
		}
		return stack;
	}
	public static ArrayList doAnd(ArrayList stack) {
		if(stack.size()<2){
            stack.add(0,":error:");
		}
		else if (((stack.get(1) == ":true:") || (stack.get(1) == ":false:"))  && ((stack.get(0) == ":true:") || (stack.get(0) == ":false:"))){
			boolean bx = false;
			boolean by = false;
			String x = (String) stack.get(1);
			String y = (String) stack.get(0);
            stack.remove(0);
            stack.remove(0);
            if(x == ":true:"){
            	bx = true;
            }
            if(y == ":true:"){
            	by = true;
            }
            if(x == ":false:"){
            	bx = false;
            }
            if(y == ":false:"){
            	by = false;
            }
            boolean newTop = bx&&by;
            if(newTop == true){
            	stack.add(0, ":true:");
            }
            if(newTop == false){
            	stack.add(0, ":false:");
            }
		}
    	else{
    		stack.add(0,":error:");
    	}
		return stack;
	}
	public static ArrayList doOr(ArrayList stack) {
		if(stack.size()<2){
            stack.add(0,":error:");
		}
		else if (((stack.get(1) == ":true:") || (stack.get(1) == ":false:"))  && ((stack.get(0) == ":true:") || (stack.get(0) == ":false:"))){
			boolean bx = false;
			boolean by = false;
			String x = (String) stack.get(1);
			String y = (String) stack.get(0);
            stack.remove(0);
            stack.remove(0);
            if(x == ":true:"){
            	bx = true;
            }
            if(y == ":true:"){
            	by = true;
            }
            if(x == ":false:"){
            	bx = false;
            }
            if(y == ":false:"){
            	by = false;
            }
            boolean newTop = bx||by;
            if(newTop == true){
            	stack.add(0, ":true:");
            }
            if(newTop == false){
            	stack.add(0, ":false:");
            }
		}
    	else{
    		stack.add(0,":error:");
    	}
		return stack;
	}
	public static ArrayList doNot(ArrayList stack) {
		if(stack.size()<1){
            stack.add(0,":error:");
		}
		else if ((stack.get(0) == ":true:") || (stack.get(0) == ":false:")){
			boolean bx = false;
			String x = (String) stack.get(0);
            stack.remove(0);
            if(x == ":true:"){
            	bx = true;
            }
            if(x == ":false:"){
            	bx = false;
            }
            boolean newTop = !bx;
            if(newTop == true){
            	stack.add(0, ":true:");
            }
            if(newTop == false){
            	stack.add(0, ":false:");
            }
		}
    	else{
    		stack.add(0,":error:");
    	}
		return stack;
	}
	public static ArrayList doEqual(ArrayList stack) {
		if(stack.size()<2){
            stack.add(0,":error:");
		}
		return stack;
	}
	public static ArrayList doLessThan(ArrayList stack) {
		return stack;
	}
	public static ArrayList doBind(ArrayList stack) {
		return stack;
	}
	public static ArrayList doIf(ArrayList stack) {
		return stack;
	}
}
